### PR TITLE
listpays

### PR DESCRIPTION
**Pull request by rustyrussell**
Also cleans up our badly-misunderstood 'description' parameter for sendpay/pay to do what people (including me!) thought it did.

@cdecker Q on what you want to do with compat in pylightning: I didn't, but we can.

### CODE REVIEW

File: contrib/pylightning/lightning/lightning.py [hunk 1]
```diff
@@ -464,18 +464,16 @@ def waitsendpay(self, payment_hash, timeout=None):
         }
         return self.call("waitsendpay", payload)
 
-    def pay(self, bolt11, msatoshi=None, description=None, riskfactor=None):
+    def pay(self, bolt11, msatoshi=None, label=None, riskfactor=None):
```
Comment (cdecker) [inline contrib/pylightning/lightning/lightning.py]: I'd propose to just add both `label` and `description` and then check that only one is set when constructing the payload (assigning that one to `label`). We can also warn about `description` being phased out and remove it once we remove the first wave of deprecated APIs.

File: plugins/pay.c [hunk 2]
```diff
@@ -1093,6 +1094,90 @@ static struct command_result *handle_paystatus(struct command *cmd,
...
+			copy_member(&ret, buf, t, "payment_preimage", ",");
+		copy_member(&ret, buf, t, "label", ",");
+		copy_member(&ret, buf, t, "amount_sent_msat", "");
+		tal_append_fmt(&ret, "}");
+	}
+	tal_append_fmt(&ret, "] }");
+	return command_success(cmd, ret);
+}
+
+static struct command_result *handle_listpays(struct command *cmd,
```
Comment (cdecker) [inline plugins/pay.c]: So far we had the convention of calling these `json_<command_name>` which I found very useful for grepping them.

File: plugins/pay.c [hunk 3]
```diff
@@ -1093,6 +1094,90 @@ static struct command_result *handle_paystatus(struct command *cmd,
...
+	char *ret;
+	bool some = false;
+
+	arr = json_get_member(buf, result, "payments");
+	if (!arr || arr->type != JSMN_ARRAY)
+		return command_fail(cmd, LIGHTNINGD,
+				    "Unexpected non-array result from listpayments");
+
+	ret = tal_fmt(cmd, "{ 'pays': [");
+	json_for_each_arr(i, t, arr) {
```
Comment (cdecker) [inline plugins/pay.c]: Am I correct in thinking that this loop will eventually coalesce all parts of a multi-part payment?

File: contrib/pylightning/lightning/lightning.py (hunk 1 above)
Comment (rustyrussell) [reply contrib/pylightning/lightning/lightning.py]: Well, if we accept both, lightningd already does that check?

File: plugins/pay.c (hunk 2 above)
Comment (rustyrussell) [reply plugins/pay.c]: Yeah, good idea.

File: plugins/pay.c (hunk 3 above)
Comment (rustyrussell) [reply plugins/pay.c]: Yes.

File: contrib/pylightning/lightning/lightning.py (hunk 1 above)
Comment (cdecker) [reply contrib/pylightning/lightning/lightning.py]: Right :+1:

**Comment by cdecker** [comment]
ACK https://github.com/ElementsProject/lightning/pull/2382/commits/a72782cc7d72c982f49a0214af5cc85101d58a26

**Pull request merged**